rust/src/treefile.rs

```diff
@@ -4668,3 +4670,82 @@ pub(crate) fn treefile_delete_client_etc() -> CxxResult<u32> {
     }
     Ok(n)
 }
+
+/// Apply a treefile to the running environment (usually during an image build)
+#[derive(Debug, Parser)]
+pub(crate) struct TreefileApplyOpts {
+    /// Path to the treefile.
+    #[clap(value_parser)]
+    treefile: Utf8PathBuf,
+}
+
+impl TreefileApplyOpts {
+    pub(crate) fn run(self) -> Result<()> {
+        let mut tf = Treefile::new_boxed(&self.treefile, Some(&utils::get_rpm_basearch()))
+            .context("parsing treefile")?;
+
+        // we only handle a subset of fields here
+        if let Some(packages) = &tf.parsed.packages {
+            run_dnf("install", &tf, packages.iter()).context("installing packages with dnf")?;
+        };
+
+        // handle postprocess scripts
+        let rootfs_dfd =
+            Dir::open_ambient_dir("/", cap_std::ambient_authority()).context("opening /")?;
+        compose_postprocess_scripts(&rootfs_dfd, &mut tf, crate::PostprocessBwrap::None)
+            .context("running postprocess scripts")?;
+
+        Ok(())
+    }
+}
+
+fn run_dnf(
+    command: &str,
+    treefile: &Treefile,
+    args: impl Iterator<Item: AsRef<OsStr>>,
+) -> Result<()> {
+    let mut cmd = Command::new("dnf");
+    cmd.arg(command).args(args);
+    cmd.arg("--noplugins");
+    cmd.arg("-y");
+
+    match treefile.parsed.documentation {
```

Comment:
It'd seem to me to be quite unusual to override the documentation flag in a *derived* build; isn't a feature here to inherit the base image's `/etc/dnf.conf`?

Maybe just error out if the option is set?

Reply:
If a user wants to inherit, they can leave `documentation` unspecified. If they want documentation for their layered packages even if the base image didn't have them, I feel like we should allow that? There's nothing technically that prevents it.

Reply:
> If they want documentation for their layered packages even if the base image didn't have them, I feel like we should allow that? There's nothing technically that prevents it.

I agree technically it's possible but I feel like it's at best a really unusual situation we could not care about or try to test - the 95% case I suspect is folks who would want to take a base image for which documentation is turned off and ensure that documentation is turned on for the base image and all the layered stuff too.

That'd be doable, in theory if we reinstall those base packages with tsflags=docs (i.e. !nodocs) we should end up in mostly the same state. Except of course for packages where we do file remapping like selinux.

Reply:
But if you want to keep this as is, that's also OK by me and if so feel free to just mark this conversation as resolved!

Reply:
WFM. We can delay until there's a concrete ask for it.

Reply:
Instead of making it a hard error, I switched it to be a warning to make it easier to use with treefiles actively used for base composes as well.

Reply:
Fine by me but wouldn't it be better to split up those two treefiles though?